### PR TITLE
Unified driver population control (partial)

### PR DESCRIPTION
**Pull request by PDoakORNL**
Traveling this weekend.  Want to get this in there before I get in car, more commits later.  Currently not compiling but I think the outline is clear. Expect to have in a running if not correct state soon.

### CODE REVIEW

File: src/QMCDrivers/MCPopulation.cpp [hunk 1]
```diff
@@ -122,11 +136,49 @@ void MCPopulation::createWalkers(IndexType num_walkers)
...
+      createWalkerInplace(*it_walkers);
+      allocateWalkerStuffInplace(walker_index);
+      break;
+    }
+    ++walker_index;
+  }
+  // you need to add to all the vectors if no walkers have been killed
+  // for now we throw since this must be implemented
+  if (walker_index == walkers_.size())
+    throw std::runtime_error("Needed to add to population in MCPopulation but it is not implemented.");
```
Comment (prckent) [inline src/QMCDrivers/MCPopulation.cpp]: Reminder to implement this before merging

**Comment by prckent** [comment]
Note the reported bug https://github.com/QMCPACK/qmcpack/issues/1817

File: src/QMCDrivers/DMC/WalkerControlMPI.cpp [hunk 2]
```diff
@@ -390,4 +447,163 @@ void WalkerControlMPI::swapWalkersSimple(MCWalkerConfiguration& W)
...
+  UPtrVector<MCPWalker> new_walkers;
+  std::vector<int> copies_of_each_new_walker;
+
+  // sort good walkers by the number of copies
+  std::vector<std::pair<int, std::reference_wrapper<MCPWalker>>> sorted_good_walkers;
+  for (int iw = 0; iw < ncopy_w.size(); iw++)
+    sorted_good_walkers.push_back(std::make_pair(adjust.copies_to_make[iw], adjust.good_walkers[iw]));
+  // Sort only on the number of copies
+  std::sort(sorted_good_walkers.begin(), sorted_good_walkers.end(),
+            [](auto& a, auto& b) { return a.first > b.first; });
```
Comment (ye-luo) [inline src/QMCDrivers/DMC/WalkerControlMPI.cpp]: I thought std::sort works with vector pairs by the first value.

File: src/QMCDrivers/DMC/WalkerControlMPI.cpp [hunk 3]
```diff
@@ -390,4 +447,163 @@ void WalkerControlMPI::swapWalkersSimple(MCWalkerConfiguration& W)
...
+ *
+ *  2005 called and asked for its synchronous messaging back
+ *  I think MPI async is more than two years old
+ *
+ *  MCPopulation is sufficiently different from MCWalkerConfiguration that this is 
+ *  basically a rewrite.  I have not replicated the more ridiculous elements of the original.
+ *  For example:
+ *   * Sending duplicates to the same reciever rank but resorting good walkers if send-receive pair 
+ *     changes.
+ *   * Sending a tiny separate blocking message to tell the receiver how many copies
```
Comment (ye-luo) [inline src/QMCDrivers/DMC/WalkerControlMPI.cpp]: This is intended to avoid sending duplicates to the same receiver. The two items mentioned here are contradictory to each other.

File: src/QMCDrivers/DMC/WalkerControlMPI.cpp [hunk 4]
```diff
@@ -390,4 +447,163 @@ void WalkerControlMPI::swapWalkersSimple(MCWalkerConfiguration& W)
...
+      WalkerMessage message = message_queue.front();
+      message_queue.pop();
+      send_requests.push_back(myComm->comm.isend_value(message.multiplicity, message.target_rank));
+      size_t byteSize    = message.walker.byteSize();
+      send_requests.push_back(myComm->comm.isend_n(message.walker.DataSet.data(), byteSize, message.target_rank));
+    }
+  }                                                    
+
+  std::vector<mpi3::request> receive_requests;
+  messages_incoming *= 2; // extra message sent for multiplicity
```
Comment (ye-luo) [inline src/QMCDrivers/DMC/WalkerControlMPI.cpp]: What is the magic number 2 here?

Comment (PDoakORNL) [reply src/QMCDrivers/DMC/WalkerControlMPI.cpp]: That was for the separate number of copies message, the next push to this PR will remove this.

File: src/QMCDrivers/DMC/WalkerControlMPI.cpp (hunk 2 above)
Comment (PDoakORNL) [reply src/QMCDrivers/DMC/WalkerControlMPI.cpp]: https://en.cppreference.com/w/cpp/utility/pair/operator_cmp
TLDR
First compare lhs.first rhs.first if lhs.first == rhs.first then compare lhs.second, rhs.second.

I considered this oversorting since from my naive perspective sorting by walker IDs seemed if anything a potential bias.
However as concurrency, threaded and MPI, cause divergence this could add to it.

So while subtle this is a change.

**Comment by ye-luo** [comment]
@prckent regarding your last point. The current production code does send out per-node weight and energy with allreduce. Because it is per-node and to avoid sending duplicated walkers, there is a small blocking near-neighbour send/receive to communicate info about walker duplication before non-blocking send/receive heavy walker data.

File: tests/solids/diamondC_2x1x1_pp/CMakeLists.txt [hunk 5]
```diff
@@ -60,6 +60,15 @@ ENDIF()
                     1 DIAMOND2_DMC_SCALARS # DMC
                     )
 
+  QMC_RUN_AND_CHECK(short-diamondC_2x1x1_pp-dmcbatch_sdj
+                    "${CMAKE_SOURCE_DIR}/tests/solids/diamondC_2x1x1_pp"
+                    qmc_short
+                    qmc_short_dmcbatch.in.xml
```
Comment (ye-luo) [inline tests/solids/diamondC_2x1x1_pp/CMakeLists.txt]: This test uses the reference values of qmc_short_vmc_dmc.in.xml but input parameters are very different. Need a fix later.

Comment (PDoakORNL) [reply tests/solids/diamondC_2x1x1_pp/CMakeLists.txt]: The test is definitely invalid and I expect it to fail.  But it doesn't crash and produces scalar.dat output.  Population control is suppressed by forcing the no_branch call on WalkerControl.  Either before or after this PR is merged that is the next task for batched DMC.

**Comment by PDoakORNL** [comment]
Their should be no significant change to the existing production code.  If the FOM can be run without population control this PR would be sufficient for DMCBatched. At this point I think we should review to merge for the small refactors to objects shared with the original drivers and limit divergence.  I agree that we should when "the dust settles" does a more holistic review of the Unified Driver.  The new population control code is not yet correct beyond the unit tests and I intend to be the subject of the next PR.

**Comment by prckent** [comment]
Quick comment on the batched tests diamondC_2x1x1_pp, dmc & possibly also vmc. Needs to be fixed when the population control is fixed and the DMC batched driver is claimed to be correct:

The batched tests should be essentially exact copies of the existing short tests except with (presumably) better statistics so that the batching is meaningfully tested and there is still sufficient equilibration. The expected statistical error can be obtained from the data we already have for the long tests.

The current dmc_batch input specifies subSteps which is not meaningful in DMC. I would prefer that we ban the specification of this value in the DMC input block; at least values !=1 should be an abort since it signifies that the user does not know what they are doing :-)

**Comment by PDoakORNL** [comment]
> specification of this value in the DMC input block; at least values !=1 should be an abort since it signifies that the user does not know what they are doing :-)

I will make the input and possibly check less rubbish when I figure out what is wrong with gcc and the new driver tests in the complex mixed precision case.

**Comment by prckent** [comment]
Also happy for this to be merged as soon as the CI passes.

**Comment by PDoakORNL** [comment]
The DMC substeps check did not make it in but I added it to my issues list.  Will be in next Unified Driver PR
